Add notification group endpoints

deepak-sreekumar opened the pull request:
### What kind of change does this PR introduce? 
Added Notification group endpoints
- GET /v1/notification-groups/:id
- PUT /v1/notification-groups/:id
- DELETE /v1/notification-groups/:id

- [ ] Bug
- [x] Feature
- [ ] Docs
- [ ] Other(s)


### Why was this change needed?
#1321 


### Other information (Screenshots)

<!-- Add notes or any other information here -->
<!-- Also add all the screenshots which support your changes -->

florian-lefebvre commented:
I wish I could review this but I am not familiar enough with the api codebase cc @scopsy

p-fernandez commented on apps/api/src/app/notification-groups/usecases/update-notification-group/update-notification-group.usecase.ts:
```diff
@@ -0,0 +1,29 @@
+      _organizationId: command.organizationId,
+    });
+
+    await this.notificationGroupRepository.update(
+      {
+        _id: command.id,
+      },
+      {
+        $set: {
+          name: command.name,
+        },
+      }
+    );
```
Ok, I just checked the implementation of `update` in the Base repository and there is the need of fetch the entity.
https://github.com/novuhq/novu/blob/e345d493d71b370e1b75485b44f33a52e50f1a7e/libs/dal/src/repositories/base-repository.ts

p-fernandez commented on apps/api/src/app/notification-groups/usecases/update-notification-group/update-notification-group.command.ts:
```diff
@@ -0,0 +1,10 @@
+import { EnvironmentWithUserCommand } from '../../../shared/commands/project.command';
+
+export class UpdateNotificationGroupCommand extends EnvironmentWithUserCommand {
+  @IsString()
```
This param should be mandatory. I am afraid of the side effects of it as not sure if not having an id could potentially make the repository to update all entries with the same name. That's something we can check with a test to know if we would need to have more safety barriers.

p-fernandez commented on apps/api/src/app/notification-groups/notification-groups.controller.ts:
```diff
@@ -44,6 +64,30 @@ export class NotificationGroupsController {
     );
   }
 
+  @Put('/:id')
+  @ExternalApiAccessible()
+  @ApiCreatedResponse({
+    type: NotificationGroupResponseDto,
+  })
+  @ApiOperation({
+    summary: 'Update notification group',
+  })
+  updateNotificationGroup(
+    @UserSession() user: IJwtPayload,
+    @Param('id') id: string,
+    @Body() body: CreateNotificationGroupRequestDto
+  ): Promise<NotificationGroupResponseDto> {
```
You are using a PUT HTTP method in the endpoint definition but you have a PATCH method implementation as you only send the fields of the entity you want to change.
So I would suggest you to either one of both:
- you change the implementation of this endpoint to expect and take the full entity to be updated in the database, or
- you modify the method of the endpoint to PATCH and leave the current implementation your did.

p-fernandez reviewed (changes requested):
Thanks for this work. Let's do some changes to have a great PR to merge. 💪🏻

deepak-sreekumar commented:
Thanks for the review @p-fernandez ! I'll try and resolve the comments In a day or two.

davidsoderberg commented on apps/api/src/app/notification-groups/usecases/delete-notification-group/delete-notification-group.usecase.ts:
```diff
@@ -0,0 +1,29 @@
+
+  async execute(command: DeleteNotificationGroupCommand) {
+    try {
+      const group = await this.notificationGroupRepository.findOne({
+        _organizationId: command.organizationId,
+      });
```
No we need to throw `NotFoundException` in those cases 😄

LetItRock commented on apps/api/src/app/notification-groups/usecases/update-notification-group/update-notification-group.usecase.ts:
```diff
@@ -0,0 +1,29 @@
+      }
+    );
+
+    const item = this.notificationGroupRepository.findById(command.id);
```
this additional variable is unnecessary

deepak-sreekumar replied:
Removed. Thanks!

scopsy commented:
Hi @deepak-sreekumar did you had a chance to take a look at @p-fernandez comments here?

deepak-sreekumar commented:
Hi @scopsy, Really sorry but I did not get a chance to address the comments yet! Let me try to spend some time on this over this weekend.

florian-lefebvre commented:
Hey @deepak-sreekumar any news on this?

deepak-sreekumar commented on apps/api/src/app/notification-groups/usecases/delete-notification-group/delete-notification-group.command.ts:
```diff
@@ -0,0 +1,7 @@
+import { IsString } from 'class-validator';
+import { EnvironmentWithUserCommand } from '../../../shared/commands/project.command';
+
+export class DeleteNotificationGroupCommand extends EnvironmentWithUserCommand {
+  @IsString()
+  id: string;
+}
```
I'm not that well versed in Nest, but Isn't the param become mandatory with `IsString` ? Tested this out and looks to be returning 400 when it's not provided. Please let me know/point me toward the decorator you have in mind. Thanks!

deepak-sreekumar commented on apps/api/src/app/notification-groups/e2e/get-notification-group.e2e.ts:
```diff
@@ -0,0 +1,27 @@
+    await session.initialize();
+  });
+
+  it('should get the notification group by id', async function () {
```
Done! Thanks for the reminder

deepak-sreekumar commented on apps/api/src/app/notification-groups/usecases/get-notification-group/get-notification-group.command.ts:
```diff
@@ -0,0 +1,7 @@
+
+export class GetNotificationGroupCommand extends EnvironmentWithUserCommand {
+  @IsString()
+  id: string;
```
Please check the query here: https://github.com/novuhq/novu/pull/1921/files#r1039794209

deepak-sreekumar commented on apps/api/src/app/notification-groups/usecases/index.ts:
```diff
@@ -1,8 +1,14 @@
   CreateNotificationGroup,
+  DeleteNotificationGroup,
+  UpdateNotificationGroup,
   //
```
Done

deepak-sreekumar commented on apps/api/src/app/notification-groups/usecases/get-notification-group/get-notification-group.usecase.ts:
```diff
@@ -0,0 +1,12 @@
+  constructor(private notificationGroupRepository: NotificationGroupRepository) {}
+
+  async execute(command: GetNotificationGroupCommand): Promise<NotificationGroupEntity> {
+    return await this.notificationGroupRepository.findById(command.id);
```
Done! Thanks

deepak-sreekumar commented:
@scopsy / @p-fernandez  Addressed the PR comments. Please take a look when you get time! 
Really sorry about the delay here. Got caught up with some urgent stuff. (This might be the case for next month as well so I'll most likely won't be able to address the comments right away but I'll try my best! :( .  )

scopsy commented on apps/api/src/app/notification-groups/usecases/delete-notification-group/delete-notification-group.usecase.ts:
```diff
@@ -0,0 +1,32 @@
+
+  async execute(command: DeleteNotificationGroupCommand) {
+    try {
+      const group = await this.notificationGroupRepository.findOne({
+        _organizationId: command.organizationId,
+      });
```
I'm a bit confused of whule this query is needed, it will return random parentId's each time. I think it's better to remove it altogether from here and only delete using `command.id` not sure that parentId should be factored in

scopsy commented on apps/api/src/app/notification-groups/usecases/get-notification-group/get-notification-group.usecase.ts:
```diff
@@ -0,0 +1,15 @@
+  constructor(private notificationGroupRepository: NotificationGroupRepository) {}
+
+  async execute(command: GetNotificationGroupCommand): Promise<NotificationGroupEntity> {
+    const result = await this.notificationGroupRepository.findById(command.id);
+    if (result === null) throw new NotFoundException();
```
Novu is a multi-tenant system hence we are required to use the `command.organizationId` or `command.environmentId` in the dal queries. Could you please update the query here.

scopsy reviewed (changes requested):
Looks good @deepak-sreekumar i have two final important things to note in the current implementation and we can merge this!

jainpawan21 commented:
@deepak-sreekumar do you need help in resolving PR comments.
Feel free to ask us in discord

jainpawan21 commented:
Closing this PR due to inactivity and new [PR ](https://github.com/novuhq/novu/pull/3274)by @peoray
